composer/callbacks/generate.py

```diff
@@ -0,0 +1,116 @@
+            saved. If an integer, it will be assumed to be in :attr:`.TimeUnit.EPOCH`.
+            Otherwise, the unit must be either :attr:`.TimeUnit.EPOCH`, :attr:`.TimeUnit.BATCH`,
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        batch_size (int): Size of a prompt batch for generation.
+        kwargs: All kwargs well be passed along to the call to generate. This is for things like `do_sample`, `top_p`, etc
+    """
+
+    def __init__(self, prompts: List[str], interval: Union[str, int, Time], batch_size: int, **kwargs: Any):
+        self.prompts = prompts
+        self.generate_kwargs = kwargs
+        self.batch_size = batch_size
+        self.check_interval = create_interval_scheduler(interval, include_end_of_training=False)
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        if state.get_elapsed_duration() is not None and self.check_interval(state, event):
+            self.generate(state, logger)
+        else:
+            super().run_event(event, state, logger)
+
+    def generate(self, state: State, logger: Logger):
+        model = state.model
+        if isinstance(model, DistributedDataParallel):
+            # Why does DistributedDataParallel wrap the HuggingFaceModel and not the other way around?
+            # Can we make this assumption?
+            model = model.module
+        assert isinstance(
+            model, HuggingFaceModel
+        ), f'Expected HuggingFaceModel, but got {state.model.__class__}'  # TODO: Extend to support any models that have a generate method.
+
+        tokenizer = model.tokenizer
+
+        try:
+            from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='nlp',
+                                                conda_package='transformers',
+                                                conda_channel='conda-forge') from e
+        Tokenizer = Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+        tokenizer = cast(Tokenizer, tokenizer)
+
+        # Set to evaluation mode and stash the original mode.
+        original_mode = model.training
+        model.eval()
+        device = state.device
+
+        # Stash the original value of padding_side because generation requires left padding
+        original_padding_side = tokenizer.padding_side
+        tokenizer.padding_side = 'left'
+        if tokenizer.pad_token_id is None:
+            tokenizer.pad_token_id = tokenizer.eos_token_id
+        tokenized_input = tokenizer(self.prompts, return_tensors='pt', padding=True)
+
+        for k, v in tokenized_input.items():
+            tokenized_input[k] = device.tensor_to_device(v)
+
+        all_input_ids = tokenized_input['input_ids']
+        all_attn_masks = tokenized_input['attention_mask']
+
+        batches = DataLoader(TensorDataset(all_input_ids, all_attn_masks), self.batch_size)
```

Comment:
Yeah, on the fence also; just thought it could be a little cleaner this way + maybe there's under-the-hood benefits of using pytorch api (though there doesn't really seem to be after looking at the impl)

I can switch over to doing a for loop. I think creating dataset/dataloader objects can be a bit overkill 😵‍💫